File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m=len(grid)
        n=len(grid[0])
        def solve(i,j):
            if D.get((i,j))!=None:
                return D[(i,j)]
            if i==m-1 and j==n-1:
                return grid[i][j]
            else:
                a=b=math.inf
                if i<m-1:
                    a=solve(i+1,j)
                if j<n-1:
                    b=solve(i,j+1)
                D[(i,j)]=min(a,b)+grid[i][j]
                return D[(i,j)]
        D={}
        return solve(0,0)

File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        for(int i=0;i<grid.size();i++)
        {
            for(int j=0;j<grid[0].size();j++)
            {
                if(i==0)
                {
                    if(j>0)grid[i][j]+=grid[i][j-1];
                    continue;
                }
                if(j==0)
                {
                    grid[i][j]+=grid[i-1][0];
                }
                else
                {
                    grid[i][j]+=min(grid[i-1][j],grid[i][j-1]);
                }
            }
        }
        return grid[grid.size()-1][grid[0].size()-1];
    }
};